### ISSUE TITLE
You cannot use unseen label info in the training process

### ISSUE THREAD

**Issue by xmyqsh** [valeoai/zero_shot_semantic_segmentation#4]
Thanks for sharing this repo.
The idea is creative. But there is some logic error in the implementation process.
In a word, you have used the unseen label info in the training process.
1. [has_unseen_class](https://github.com/valeoai/ZS3/blob/master/zs3/train_pascal_GMMN.py#L204): You don't know you have unseen class or not in real world. You just know the whole set of classes, classes set with label and classes set without label.
You should use `has_unseen_class = len(unique_class) < 21` for VOC.
2. [unique_class](https://github.com/valeoai/ZS3/blob/master/zs3/train_pascal_GMMN.py#L205): You don't know the `unique_class` of a image in real world. You just know the unique labeled classes in that image, and the whole set of unseen classes, and the whole set of classes.
You should generate all classes, seen and unseen, here. And use the unique seen classes in a image to train your generator.
3. [loss = self.criterion(output, target)](https://github.com/valeoai/ZS3/blob/master/zs3/train_pascal_GMMN.py#L265): You cannot use the virtual unseen label in the target to train your classifier. It is not fair. You should add some lines of code:
`
for unseen_label in whole_unseen_label:
    target[target==unseen_label] = 255
`
And set the `ignore_label` parameter in `CEloss` to `255` to ignore the virtual unseen label to simulate the real world application totally.

Despite all of above, I still believe your idea works. Hope to see your updated result.

**Comment by maximebucher** [comment]
Hello xmyqsh,

thank you for pointing this issue.

In order to make the implementation "real world compatible", you can deduce the statistics of the unseen classes from the seen classes according to the word2vec distance. By doing this, the performance is quite similar.

I will soon provide an updated version.

Maxime

**Comment by tuanhungvu** [comment]
https://github.com/valeoai/ZS3/issues/10#issuecomment-740829384
